json.h

```diff
@@ -15,7 +15,7 @@ struct json_value {
 	union {
 		long long integer_number;
 		double float_number;
-		char *string;
+		char const *string;
 		struct json_object *object;
 		struct json_array *array;
 	};
```

Comment:
I suggest having one more extra member for the "JSON add method" purpose, here we can have `char const *const_str;` instead of changing `char *string`, so it's still compatible with other parts that are using it.

 
```
	union {
		long long integer_number;
		double float_number;
		char *string;
		char const *const_str;
		struct json_object *object;
		struct json_array *array;
	};
```

And then to avoid `strdup`, you can assign to `const_str`.

Reply:
As you might have noticed there has been a `type` field designating what a field in the union should be used. How are you going to distinguish between `string` and `const string`? Do you want to offer a new enum item for the field `type`?

Reply:
@axboe Actually, I cannot understand why you have denied my changes about freeing string that has been duplicated. `strdup` is used there because there is a mutable string in `struct json_value`, and this is a part of this json library inside FIO. If I remove `strdup` I cannot assign immutable string to mutable, you know, and casting away `const` looks to me illegal, but if you insist on doing so, OK, we can do this ugly approach. On the other hand @ammarfaizi2 offers to introduce a new field `const string`, what looks as a workaround for the sake of workaround, although it is acceptable, but the latter requires to introduce a new enumeration item for the field `type` in `struct json_value`. We need your decision then

Reply:
> As you might have noticed there has been a `type` field designating what a field in the union should be used. How are you going to distinguish between `string` and `const string`? Do you want to offer a new enum item for the field `type`?

OK, if you follow that that `json_object_add_value_type()` with argument type string, it doesn't actually change anything, so the `strdup()` isn't necessary, I suggest the following changes:
```diff
diff --git a/json.h b/json.h
index d9824263..ec87751b 100644
--- a/json.h
+++ b/json.h
@@ -81,8 +81,13 @@ static inline int json_object_add_value_string(struct json_object *obj,
 	struct json_value arg = {
 		.type = JSON_TYPE_STRING,
 	};
+	union {
+		const char *a;
+		char *b;
+	} p;
 
-	arg.string = strdup(val ? : "");
+	p.a = val ? val : "";
+	arg.string = p.b;
 	return json_object_add_value_type(obj, name, &arg);
 }
 

```
which is much simpler and faster.

Previously, you said it's illegal, but it's actually not. Because the callee doesn't change anything. Comment?

Reply:
I tend to agree to accept this snippet of code. My point has been to get rid of memory leak in JSON